wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/modal/ModalDialog.java

```diff
@@ -0,0 +1,240 @@
+package org.apache.wicket.extensions.ajax.markup.html.modal;
+
+import java.io.Serializable;
+
+import org.apache.wicket.Application;
+import org.apache.wicket.ajax.AjaxRequestTarget;
+import org.apache.wicket.markup.head.IHeaderResponse;
+import org.apache.wicket.markup.head.JavaScriptHeaderItem;
+import org.apache.wicket.markup.head.OnDomReadyHeaderItem;
+import org.apache.wicket.markup.html.WebMarkupContainer;
+import org.apache.wicket.markup.html.form.Form;
+import org.apache.wicket.markup.html.panel.EmptyPanel;
+import org.apache.wicket.markup.html.panel.Panel;
+
+import com.github.openjson.JSONStringer;
+
+/**
+ * Presents a modal dialog to the user. See open and close methods.
+ * 
+ * @author Igor Vaynberg (ivaynberg)
+ */
+public class ModalDialog extends Panel
+{
+
+	public static final String CONTENT_ID = "content";
+
+	private final WebMarkupContainer container;
+	private final WebMarkupContainer contentContainer;
+	private boolean open = false;
+	private transient boolean openedInThisRequest = false;
+	private Options options;
+
+	public ModalDialog(String id)
+	{
+		super(id);
+		setOutputMarkupId(true);
+
+		// Container controls the overall visibility of the modal form innards. In its initial state
+		// this is set to
+		// hidden so only the external tag renders. When the modal is openedInThisRequest it is
+		// first repainted with
+		// this shown to
+		// insert the markup into the dom, and then the modal javascript is invoked to rip out this
+		// tag out of the dom
+		// and make it modal.
+		container = new WebMarkupContainer("container");
+		container.setOutputMarkupId(true);
+		container.setVisible(false);
+		add(container);
+
+		// We need this here in case the modal itself is placed inside a form. If that is the case
```

Comment:
can we do this in `onInitialize()` and add it only if there is / is not an outer form ?

Reply:
what would you gain? if you are not in a form you would have to add a webmarkupcontainer instead with overridden oncompoennttag that mutates the form tag to a div in markup. it seems like needless complexity to me...

Reply:
Actually the question should be "Why do we need a Form here at all ?"
Why not let the application provide a Form inside the `content` container if it needs one or use the outer (if there is such) ? 
I am on my mobile now, so I cannot re-read ModalWindow's javadoc. But it would be good to simplify here as much as possible. AFAIR Wicket-Bootstrap's Modal component does not provide such mandatory Form and no one complained.

Reply:
suppose
* form(1) -> modal -> content -> form(2) -> submit
* because modal is a child of form(1) form(2) will render its tag as a div because html does not support nested forms
* then it is moved into its own direct child of body by modal js
* so now submit button does not have a form tag to work with